sky/server/requests/executor.py

```diff
@@ -343,112 +444,88 @@ def enqueue():
+    - Local mode (deploy=False), we assume the server is running in a shared
+      environment (e.g. laptop) and users typically do not pay attention to
+      the resource setup of the server. Moreover, existing users may expect
+      some consistent behaviors with old versions, i.e. before API server was
+      introduced, so:
+      - The max number of long-running executor processes are limited, to avoid
+        high memory consumption when the server is idle.
+      - Allow burstable workers to handle requests when all long-running
+        workers are busy, which mimics the behavior of local sky CLI before
+        API server was introduced.
+      - Works in low resources environments, and further reduce the memory
+        consumption in low resource environments.
+
+    Note that there is still significant overhead for SDK users when migrate to
+    local API server. Since the users are free to run sky operations in Threads
+    when using SDK but all client operations will occupy at least one worker
+    process after API server was introduced.
+    """
     # Determine the job capacity of the workers based on the system resources.
     cpu_count = common_utils.get_cpu_count()
     mem_size_gb = common_utils.get_mem_size_gb()
     mem_size_gb = max(0, mem_size_gb - server_constants.MIN_AVAIL_MEM_GB)
+    # Runs in low resource mode if the available memory is less than
+    # server_constants.MIN_AVAIL_MEM_GB.
     max_parallel_for_long = _max_long_worker_parallism(cpu_count,
                                                        mem_size_gb,
                                                        local=not deploy)
     max_parallel_for_short = _max_short_worker_parallism(
         mem_size_gb, max_parallel_for_long)
-    logger.info(
-        f'SkyPilot API server will start {max_parallel_for_long} workers for '
-        f'long requests and will allow at max '
-        f'{max_parallel_for_short} short requests in parallel.')
+    if mem_size_gb < server_constants.MIN_AVAIL_MEM_GB:
+        if deploy:
+            # For deployment, we require at the min available memory to be
+            # met for stable operation.
+            raise RuntimeError(
+                f'Not enough memory to deploy the API server. '
+                f'{mem_size_gb}GB available, '
+                f'but {server_constants.MIN_AVAIL_MEM_GB}GB required.')
+        else:
+            # Permanent worker process may have significant memory consumption
+            # (~350MB per worker) after running commands like `sky check`, so we
+            # don't start any permanent workers in low resource local mode. This
+            # mimics the behavior of local sky CLI before API server was
+            # introduced, where the CLI will start new process everytime and
+            # never reject to start due to resource constraints.
+            # Note that the refresh daemon will still occupy one worker
+            # permanently because it never exits.
```

Comment:
```suggestion
            # Permanent worker process may have significant memory consumption
            # (~350MB per worker) after running commands like `sky check`, so we
            # don't start any permanent workers in low resource local mode. This
            # mimics the behavior of local sky CLI before API server was
            # introduced, where the CLI will start new process everytime and
            # never reject to start due to resource constraints.
            # Note that the refresh daemon will still occupy one worker
            # permanently because it never exits.
            # For correctness, because the available memory is already below
            # the threshold, there will be only 1 uvicorn worker writing to the
            # queue.
```

Reply:
It seems the reason why I felt a bit weird for this part of the code is that it implicitly rely on the fact that the uvicorn will have only 1 worker when the memory is low, i.e. the handling of memory on the other part of the codebase. minor: an improvement we can do is that make the `mem_size_gb < server_constants.MIN_AVAIL_MEM_GB` comparison a common function `def is_low_memory_env()`, and have both here and the place we decide the number of workers for uvicorn.

Reply:
good catch!

Reply:
I just realized that we can simply the condition to use `LOCAL` queue when `deploy=false` (local mode), and rely on we only runs 1 unvicorn worker in local mode did not introduce new knowledge (though there is still implicit dependency, but seems okay to me)

Reply:
Nice! That probably means a lower starting overhead locally? Supposedly, the `sky api start` would be much faster?

To make sure we don't accidentally making changes to the semantic of `deploy=False`, we should:
1. Add a comment in the uvicorn worker part of code, saying that the worker count should be 1 when `deploy=False`.
2. Can we add a smoke or unit test to make sure everything added to the queue by the fastapi functions are all handled by executors in both non-deploy and deploy mode? This is to make sure that dispatcher and the uvicorn workers are connecting to the same queue.

Reply:
Sure

Reply:
Looks unit test cannot help in this case -- we will end up with something like [test_queue_dispatcher.py](https://github.com/skypilot-org/skypilot/blob/03991d0554083dea87d179fe2fc0472f5893e8ee/tests/load_tests/test_queue_dispatcher.py) that cannot verify the end to end correctness. I'm thinking of adding a flag to smoke test suite

Reply:
Added a case that relies on the `--remote-server` flag to test deploy/local modes